core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -350,15 +353,16 @@ private[kafka] object LogValidator extends Logging {
 
       try {
         for (record <- batch.asScala) {
+          val expectedOffset = expectedInnerOffset.getAndIncrement()
           if (sourceCodec != NoCompressionCodec && record.isCompressed)
             throw new InvalidRecordException("Compressed outer record should not have an inner record with a " +
-              s"compression attribute set: $record")
-          validateRecord(batch, topicPartition, record, now, timestampType, timestampDiffMaxMs, compactedTopic, brokerTopicStats)
+              s"compression attribute set: $record",
+              java.util.Collections.singletonMap(expectedOffset.toInt.asInstanceOf[java.lang.Integer], ""))
+          validateRecord(batch, topicPartition, record, expectedOffset, now, timestampType, timestampDiffMaxMs, compactedTopic, brokerTopicStats)
 
           uncompressedSizeInBytes += record.sizeInBytes()
           if (batch.magic > RecordBatch.MAGIC_VALUE_V0 && toMagic > RecordBatch.MAGIC_VALUE_V0) {
             // inner records offset should always be continuous
-            val expectedOffset = expectedInnerOffset.getAndIncrement()
             if (record.offset != expectedOffset) {
```

Comment:
I am not sure why we don't do this validation for both compressed and non-compressed batches. Seems like we should.